File: game_mechanic_mockup/mock.py
import random
import pandas as pd

# Game setup
num_players = int(input("Enter the number of players (2-8): "))  # User chooses the number of players
columns = [
    'Player Throwing', 'Die 1', 'Die 2', 'Sum', 'Instruction 1', 'Instruction 2',
    'Rule 1', 'Rule 2', 'Rule 3', 'Commands', 'Rival Chosen', 'Counterattack Decision', 'Lore'
]
results_df = pd.DataFrame(columns=columns + [f'Player {i} Sips' for i in range(1, num_players + 1)])

# User assigns "Knights of 3"
print("You must now assign the 'Knights of 3'. These players will drink every time a 3 is rolled.")
knights_of_3 = []
while True:
    knight = int(input(f"Enter player number to be a Knight of 3 (1-{num_players}, or 0 to finish): "))
    if knight == 0:
        break
    elif knight not in knights_of_3 and 1 <= knight <= num_players:
        knights_of_3.append(knight)
    else:
        print("Invalid input. Try again.")
print(f"Knights of 3: {knights_of_3}")

current_player = 1
consecutive_throws = 0
rules = [""] * 3
sips = {player: 0 for player in range(1, num_players + 1)}
knight_sips = {player: 0 for player in range(1, num_players + 1)}  # Track Knight-only sips separately
king_intro_shown = False  # Tracks if the King explanation has been shown
elf_intro_shown = False  # Tracks if the House Elf explanation has been shown

# Functions
def roll_dice():
    """Roll two dice and return their values and sum."""
    die1, die2 = random.randint(1, 6), random.randint(1, 6)
    return die1, die2, die1 + die2

def handle_sips(sips, player, amount, reason):
    """Add sips to a player and return a message."""
    sips[player] += amount
    return f"Player {player} drinks {amount} sip(s) ({reason})"

def counterattack(current_player, rival, double_chain):
    """Handle the counterattack logic."""
    decisions = []
    while True:
        decision = input(f"Player {rival}, do you want to counterattack? (yes/no): ").strip().lower()
        decisions.append(decision)
        if decision == "yes":
            print(f"\nPlayer {rival}, it's your turn to counterattack. Press Enter to roll the dice.")
            input(">> Press Enter to roll... ")  # Prompt to roll dice
            rival_die1, rival_die2, _ = roll_dice()
            print(f"Player {rival} rolls: {rival_die1}, {rival_die2}")
            
            if rival_die1 == rival_die2:  # Rival rolls doubles
                print(f"Player {rival} rolls doubles! Counter continues.")
                double_chain += rival_die1
                current_player, rival = rival, current_player  # Switch roles
            else:  # Rival fails
                penalty = double_chain + min(rival_die1, rival_die2)
                message = f"Player {rival} failed the counterattack and drinks {penalty} sip(s)."
                print(message)
                sips[rival] += penalty
                return rival, penalty, "failed counterattack", decisions
        else:  # Rival refuses to counterattack
            message = f"Player {rival} refuses to counterattack and drinks {double_chain} sip(s)."
            print(message)
            sips[rival] += double_chain
            return rival, double_chain, "refusal to counterattack", decisions

# Game loop
for _ in range(100):
    print(f"\nPlayer {current_player}, it's your turn. Press Enter to roll the dice.")
    input(">> Press Enter to roll... ")  # Prompt to roll dice

    die1, die2, sum_dice = roll_dice()
    is_double = die1 == die2
    consecutive_throws += 1
    double_chain = die1 if is_double else 0
    instruction2 = ""
    command_message = ""
    rival_chosen = None
    counterattack_decisions = []

    print(f"Player {current_player} rolls: {die1}, {die2} (Sum: {sum_dice})")

    # Determine "Knights of 3" rule
    if 3 in [die1, die2]:
        instruction2 = "all knights drink"
        for knight in knights_of_3:
            sips[knight] += 1
            knight_sips[knight] += 1  # Track Knight-only sips separately
        knight_message = f"Knights of 3 ({', '.join(map(str, knights_of_3))}) drink 1 sip."
        print(knight_message)
        command_message += knight_message

    # Sipping rules
    if sum_dice == 7:
        previous_player = current_player - 1 if current_player > 1 else num_players
        message = handle_sips(sips, previous_player, 1, "rolled a 7")
        print(message)
        command_message += f" | {message}"
    elif sum_dice == 8:
        for player in range(1, num_players + 1):
            sips[player] += 1
        print("Everybody drinks 1 sip!")
        last_player = int(input("Who was the last to put their cup down? Enter player number: "))
        sips[last_player] += 1
        message = f"Player {last_player} drinks an additional sip for being last."
        print(message)
        command_message += f" | Everybody drinks 1 sip, {message}"
    elif sum_dice == 9:
        next_player = (current_player % num_players) + 1
        message = handle_sips(sips, next_player, 1, "rolled a 9")
        print(message)
        command_message += f" | {message}"
    elif is_double:
        rival_chosen = int(input(f"Player {current_player}, choose your rival (1-{num_players}): "))
        rival, penalty, reason, counterattack_decisions = counterattack(current_player, rival_chosen, double_chain)
        message = f"Player {rival} drinks {penalty} sip(s) due to {reason}."
        command_message += f" | {message}"

    # New rule creation
    if consecutive_throws == 3 and (sum_dice in [7, 8, 9, 11, 12] or is_double):
        input_new_rule = input(f"Player {current_player}, type your new rule: ")
        rules.pop(0)
        rules.append(f"Rule: {input_new_rule}")
        print(f"New rule added: {input_new_rule}")
        instruction1 = "pass the phone"
        consecutive_throws = 0
    else:
        if (sum_dice in [7, 8, 9, 11, 12] or is_double):
            instruction1 = "throw again"
        else:
            instruction1 = "pass the phone"
            consecutive_throws = 0

    # Determine "King of the Castle" and "House Elf"
    adjusted_sips = {player: sips[player] - knight_sips[player] for player in sips}  # Exclude Knight sips
    min_sips = min(adjusted_sips.values())
    max_sips = max(adjusted_sips.values())
    king_of_the_castle = [player for player, total_sips in adjusted_sips.items() if total_sips == min_sips]
    house_elf = [player for player, total_sips in adjusted_sips.items() if total_sips == max_sips]

    # Lore description
    lore_message = ""
    if len(king_of_the_castle) == 1:
        lore_message += f"Player {king_of_the_castle[0]} is the King of the Castle. Do not make eye contact!"
        if not king_intro_shown:
            lore_message += " The King of the Castle may not be looked in the eye. If you do, you must drink a sip as punishment."
            king_intro_shown = True
        print(lore_message)
    if len(house_elf) == 1:
        lore_message += f" Player {house_elf[0]} is the House Elf. Do not let them make eye contact!"
        if not elf_intro_shown:
            lore_message += " The House Elf may not look anyone in the eye. If he dares to do so, he must drink a sip as punishment."
            elf_intro_shown = True
        print(lore_message)

    # Record round data
    data = {
        'Player Throwing': current_player,
        'Die 1': die1,
        'Die 2': die2,
        'Sum': sum_dice,
        'Instruction 1': instruction1,
        'Instruction 2': instruction2,
        'Rule 1': rules[0],
        'Rule 2': rules[1],
        'Rule 3': rules[2],
        'Commands': command_message.strip(' |'),
        'Rival Chosen': rival_chosen,
        'Counterattack Decision': " -> ".join(counterattack_decisions) if counterattack_decisions else None,
        'Lore': lore_message
    }
    for i in range(1, num_players + 1):
        data[f'Player {i} Sips'] = sips[i]

    results_df = pd.concat([results_df, pd.DataFrame([data])], ignore_index=True)

    # Move to next player if phone is passed
    if instruction1 == "pass the phone":
        current_player = (current_player % num_players) + 1

# Display results
print("\nGame Log:")
print(results_df)

File: party_app/lib/application_logic/mock.dart

/**
import 'dart:io';
import 'dart:math';

void main() {
  // Game setup
  final Random random = Random();
  stdout.write("Enter the number of players (2-8): ");
  final int numPlayers = int.parse(stdin.readLineSync()!); // User decides number of players
  print('Number of players: $numPlayers');

  // Columns for the game log
  final List<String> columns = [
    'Player Throwing',
    'Die 1',
    'Die 2',
    'Sum',
    'Instruction 1',
    'Instruction 2',
    'Rule 1',
    'Rule 2',
    'Rule 3',
    'Commands',
    'Rival Chosen',
    'Counterattack Decision',
    'Lore'
  ];

  // Assign Knights of 3
  print("You must now assign the 'Knights of 3'. These players will drink every time a 3 is rolled.");
  final List<int> knightsOf3 = [];
  while (true) {
    stdout.write("Enter player number to be a Knight of 3 (1-$numPlayers, or 0 to finish): ");
    int knight = int.parse(stdin.readLineSync()!);
    if (knight == 0) break;
    if (!knightsOf3.contains(knight) && knight >= 1 && knight <= numPlayers) {
      knightsOf3.add(knight);
    } else {
      print("Invalid input. Try again.");
    }
  }
  print('Knights of 3: $knightsOf3');

  int currentPlayer = 1;
  int consecutiveThrows = 0;
  final List<String> rules = List.filled(3, '');
  final Map<int, int> sips = {for (var i = 1; i <= numPlayers; i++) i: 0};
  final Map<int, int> knightSips = {for (var i = 1; i <= numPlayers; i++) i: 0};
  bool kingIntroShown = false;
  bool elfIntroShown = false;

  // Dice rolling function
  List<int> rollDice() {
    int die1 = random.nextInt(6) + 1;
    int die2 = random.nextInt(6) + 1;
    return [die1, die2, die1 + die2];
  }

  // Handling sips
  String handleSips(Map<int, int> sips, int player, int amount, String reason) {
    sips[player] = (sips[player] ?? 0) + amount;
    return 'Player $player drinks $amount sip(s) ($reason)';
  }

  // Counterattack logic
  Map<String, dynamic> counterattack(int currentPlayer, int rival, int doubleChain) {
    List<String> decisions = [];
    while (true) {
      stdout.write("Player $rival, do you want to counterattack? (yes/no): ");
      String decision = stdin.readLineSync()!.toLowerCase();
      decisions.add(decision);
      if (decision == "yes") {
        print("Player $rival, press Enter to roll the dice for the counterattack.");
        stdin.readLineSync();
        final dice = rollDice();
        print("Player $rival rolls: ${dice[0]}, ${dice[1]}");
        if (dice[0] == dice[1]) {
          print("Player $rival rolls doubles! Counter continues.");
          doubleChain += dice[0];
          int temp = currentPlayer;
          currentPlayer = rival;
          rival = temp;
        } else {
          int penalty = doubleChain + min(dice[0], dice[1]);
          sips[rival] = (sips[rival] ?? 0) + penalty;
          return {'rival': rival, 'penalty': penalty, 'reason': 'failed counterattack', 'decisions': decisions};
        }
      } else {
        sips[rival] = (sips[rival] ?? 0) + doubleChain;
        return {'rival': rival, 'penalty': doubleChain, 'reason': 'refusal to counterattack', 'decisions': decisions};
      }
    }
  }

  // Game loop
  for (int round = 0; round < 100; round++) {
    print("\nPlayer $currentPlayer, press Enter to roll the dice.");
    stdin.readLineSync();

    final dice = rollDice();
    int die1 = dice[0];
    int die2 = dice[1];
    int sumDice = dice[2];
    bool isDouble = die1 == die2;
    consecutiveThrows++;
    int doubleChain = isDouble ? die1 : 0;
    String instruction1 = '';
    String instruction2 = '';
    String commandMessage = '';
    String loreMessage = '';
    int? rivalChosen;
    List<String> counterattackDecisions = [];

    // Knights of 3 rule
    if (die1 == 3 || die2 == 3) {
      instruction2 = 'all knights drink';
      for (var knight in knightsOf3) {
        sips[knight] = (sips[knight] ?? 0) + 1;
        knightSips[knight] = (knightSips[knight] ?? 0) + 1;
      }
      commandMessage += 'Knights of 3 ($knightsOf3) drink 1 sip. ';
    }

    // Sipping rules
    if (sumDice == 7) {
      int previousPlayer = currentPlayer > 1 ? currentPlayer - 1 : numPlayers;
      commandMessage += handleSips(sips, previousPlayer, 1, 'rolled a 7');
    } else if (sumDice == 8) {
      for (var player in sips.keys) {
        sips[player] = (sips[player] ?? 0) + 1;
      }
      stdout.write("Who was the last to put their cup down? Enter player number: ");
      int lastPlayer = int.parse(stdin.readLineSync()!);
      sips[lastPlayer] = (sips[lastPlayer] ?? 0) + 1;
      commandMessage += 'Everybody drinks 1 sip, Player $lastPlayer drinks an additional sip.';
    } else if (sumDice == 9) {
      int nextPlayer = currentPlayer % numPlayers + 1;
      commandMessage += handleSips(sips, nextPlayer, 1, 'rolled a 9');
    } else if (isDouble) {
      stdout.write("Player $currentPlayer, choose your rival (1-$numPlayers): ");
      rivalChosen = int.parse(stdin.readLineSync()!);
      var result = counterattack(currentPlayer, rivalChosen, doubleChain);
      commandMessage += 'Player ${result['rival']} drinks ${result['penalty']} sip(s) due to ${result['reason']}.';
      counterattackDecisions = result['decisions'];
    }

    // New rule creation
    if (consecutiveThrows == 3 && (sumDice == 7 || sumDice == 8 || sumDice == 9 || sumDice == 11 || sumDice == 12 || isDouble)) {
      stdout.write("Player $currentPlayer, type your new rule: ");
      String newRule = stdin.readLineSync()!;
      rules.removeAt(0);
      rules.add(newRule);
      commandMessage += 'New rule added: $newRule.';
      instruction1 = 'pass the phone';
      consecutiveThrows = 0;
    } else {
      instruction1 = (sumDice == 7 || sumDice == 8 || sumDice == 9 || sumDice == 11 || sumDice == 12 || isDouble) ? 'throw again' : 'pass the phone';
      if (instruction1 == 'pass the phone') consecutiveThrows = 0;
    }

    // Determine "King of the Castle" and "House Elf"
    var adjustedSips = Map.from(sips);
    for (var player in knightsOf3) {
      adjustedSips[player] = (adjustedSips[player] ?? 0) - (knightSips[player] ?? 0);
    }
    int minSips = adjustedSips.values.reduce(min);
    int maxSips = adjustedSips.values.reduce(max);
    var kingOfTheCastle = adjustedSips.entries.where((entry) => entry.value == minSips).map((entry) => entry.key).toList();
    var houseElf = adjustedSips.entries.where((entry) => entry.value == maxSips).map((entry) => entry.key).toList();

    // Lore description
    if (kingOfTheCastle.length == 1) {
      loreMessage += 'Player ${kingOfTheCastle.first} is the King of the Castle. Do not make eye contact!';
      if (!kingIntroShown) {
        loreMessage += ' The King of the Castle may not be looked in the eye. If you do, you must drink a sip.';
        kingIntroShown = true;
      }
    }
    if (houseElf.length == 1) {
      loreMessage += ' Player ${houseElf.first} is the House Elf. Do not let them make eye contact!';
      if (!elfIntroShown) {
        loreMessage += ' The House Elf may not look anyone in the eye. If they do, they must drink a sip.';
        elfIntroShown = true;
      }
    }

    // Print round data
    print('\nRound ${round + 1}');
    print('Player Throwing: $currentPlayer');
    print('Dice: $die1, $die2 (Sum: $sumDice)');
    print('Commands: $commandMessage');
    print('Lore: $loreMessage');
    print('Sips: $sips');

    // Move to next player if phone is passed
    if (instruction1 == 'pass the phone') {
      currentPlayer = currentPlayer % numPlayers + 1;
    }
  }

  // End of game
  print('\nGame Over!');
}
    **/